python/ray/train/xgboost/_xgboost_utils.py

```diff
@@ -0,0 +1,194 @@
+from ray.util.annotations import PublicAPI
+
+try:
+    from xgboost.callback import TrainingCallback
+except ImportError:
+
+    class TrainingCallback:
+        pass
+
+
+class TuneCallback(TrainingCallback):
+    # TODO(justinvyu): [code_removal] Remove this after enforcing min xgboost version.
+    """Base class for Tune's XGBoost callbacks."""
+
+    def __call__(self, env):
+        """Compatibility with xgboost<1.3"""
+        return self.after_iteration(
+            env.model, env.iteration, env.evaluation_result_list
+        )
+
+    def after_iteration(self, model: Booster, epoch: int, evals_log: Dict):
+        raise NotImplementedError
+
+
+@PublicAPI(stability="beta")
+class RayTrainReportCallback(TuneCallback):
+    """XGBoost callback to save checkpoints and report metrics.
+
+    Args:
+        metrics: Metrics to report. If this is a list,
+            each item describes the metric key reported to XGBoost,
+            and it will be reported under the same name.
+            This can also be a dict of {<key-to-report>: <xgboost-metric-key>},
+            which can be used to rename xgboost default metrics.
+        filename: Customize the saved checkpoint file type by passing
+            a filename. Defaults to "model.json".
+        frequency: How often to save checkpoints, in terms of iterations.
+            Defaults to 0 (no checkpoints are saved during training).
+        checkpoint_at_end: Whether or not to save a checkpoint at the end of training.
+        results_postprocessing_fn: An optional Callable that takes in
+            the metrics dict that will be reported (after it has been flattened)
+            and returns a modified dict. For example, this can be used to
+            average results across CV fold when using ``xgboost.cv``.
+
+    Examples
+    --------
+
+    Reporting checkpoints and metrics to Ray Tune when running many
+    independent xgboost trials (without data parallelism within a trial).
+
+    .. testcode::
+        :skipif: True
+
+        import xgboost
+
+        from ray.tune import Tuner
+        from ray.train.xgboost import RayTrainReportCallback
+
+        def train_fn(config):
+            # Report log loss to Ray Tune after each validation epoch.
+            bst = xgboost.train(
+                ...,
+                callbacks=[
+                    RayTrainReportCallback(
+                        metrics={"loss": "eval-logloss"}, frequency=1
+                    )
+                ],
+            )
+
+        tuner = Tuner(train_fn)
+        results = tuner.fit()
+
+    Loading a model from a checkpoint reported by this callback.
+
+    .. testcode::
+        :skipif: True
+
+        from ray.train.xgboost import RayTrainReportCallback
+
+        # Get a `Checkpoint` object that is saved by the callback during training.
+        result = trainer.fit()
+        booster = RayTrainReportCallback.get_model(result.checkpoint)
+
+    """
+
+    CHECKPOINT_NAME = "model.json"
```

Comment:
Should we just use `ubj` format instead here since it'll be the standard for xgboost?